### PR TITLE
[AMDGPU][True16][MC] Support VOP3 only instructions with true16 and fake16

### PR DESCRIPTION
**Pull request by broxigarchen**
Update VOP3 only instructions with true16 and fake16 formats. 

This patch includes instructions:
V_MUL_LO_U16
V_MAX_U16
V_MAX_I16
V_MIN_U16
V_MIN_I16
V_LSHLREV_B16
V_LSHRREV_B16
V_ASHRREV_I16

### CODE REVIEW

**Review by kosarev — APPROVED** [review]
LGTM, provided the test changes are not far from what we have downstream, so not much of manual work will be needed to merge them there. (Otherwise, please will you make sure to cherry-pick this downstream before submitting here. Thanks.)

File: llvm/test/MC/Disassembler/AMDGPU/gfx12_dasm_vop3_dpp8.txt [hunk 1]
```diff
@@ -233,10 +233,16 @@
 # GFX12: v_and_or_b32_e64_dpp v255, v255, v255, src_scc dpp8:[0,0,0,0,0,0,0,0] fi:1 ; encoding: [0xff,0x00,0x57,0xd6,0xea,0xfe,0xf7,0x03,0xff,0x00,0x00,0x00]
 0xff,0x00,0x57,0xd6,0xea,0xfe,0xf7,0x03,0xff,0x00,0x00,0x00
```
Comment (Sisyph) [inline llvm/test/MC/Disassembler/AMDGPU/gfx12_dasm_vop3_dpp8.txt:235]: hi register tests?

**Review by Sisyph — COMMENTED** [review]
It seems to be missing .h tests for all the disassembler test files.

File: llvm/test/MC/Disassembler/AMDGPU/gfx12_dasm_vop3_dpp8.txt (hunk 1 above)
Comment (broxigarchen) [reply llvm/test/MC/Disassembler/AMDGPU/gfx12_dasm_vop3_dpp8.txt:235]: added

**Comment by broxigarchen** [comment]
The windows build is pass while the failing tests in linux build seems unrelated to this patch. Also checked some other PR and they have has similar error signature.

This patch is ready for another quick reivew. Thanks!

File: llvm/test/MC/Disassembler/AMDGPU/gfx11_dasm_vop3.txt [hunk 2]
```diff
@@ -549,49 +549,94 @@
...
+# W64-REAL16: v_ashrrev_i16 v5.l, src_scc, vcc_lo     ; encoding: [0x05,0x00,0x3a,0xd7,0xfd,0xd4,0x00,0x00]
+# W64-FAKE16: v_ashrrev_i16 v5, src_scc, vcc_lo       ; encoding: [0x05,0x00,0x3a,0xd7,0xfd,0xd4,0x00,0x00]
 0x05,0x00,0x3a,0xd7,0xfd,0xd4,0x00,0x00
 
-# GFX11: v_ashrrev_i16 v255, 0xfe0b, vcc_hi      ; encoding: [0xff,0x00,0x3a,0xd7,0xff,0xd6,0x00,0x00,0x0b,0xfe,0x00,0x00]
+# W32-REAL16: v_ashrrev_i16 v255.l, 0xfe0b, vcc_hi    ; encoding: [0xff,0x00,0x3a,0xd7,0xff,0xd6,0x00,0x00,0x0b,0xfe,0x00,0x00]
+# W32-FAKE16: v_ashrrev_i16 v255, 0xfe0b, vcc_hi      ; encoding: [0xff,0x00,0x3a,0xd7,0xff,0xd6,0x00,0x00,0x0b,0xfe,0x00,0x00]
+# W64-REAL16: v_ashrrev_i16 v255.l, 0xfe0b, vcc_hi    ; encoding: [0xff,0x00,0x3a,0xd7,0xff,0xd6,0x00,0x00,0x0b,0xfe,0x00,0x00]
+# W64-FAKE16: v_ashrrev_i16 v255, 0xfe0b, vcc_hi      ; encoding: [0xff,0x00,0x3a,0xd7,0xff,0xd6,0x00,0x00,0x0b,0xfe,0x00,0x00]
 0xff,0x00,0x3a,0xd7,0xff,0xd6,0x00,0x00,0x0b,0xfe,0x00,0x00
```
Comment (Sisyph) [inline llvm/test/MC/Disassembler/AMDGPU/gfx11_dasm_vop3.txt:641]: missing v_ashrrev_i16 .h test

**Review by Sisyph — COMMENTED** [review]
Looks like its still missing ashrrev .h tests from all disassembler files. And its missing any ashrrev .l or .h tests in the assembler files.

File: llvm/test/MC/Disassembler/AMDGPU/gfx11_dasm_vop3.txt (hunk 2 above)
Comment (broxigarchen) [reply llvm/test/MC/Disassembler/AMDGPU/gfx11_dasm_vop3.txt:641]: added. Sorry I made a typo in my script making v_assrrev_i16 to be b16 and cause this to be missing

File: llvm/test/MC/Disassembler/AMDGPU/gfx12_dasm_vop3_dpp8.txt [hunk 3]
```diff
@@ -233,12 +233,70 @@
...
+# W64-FAKE16: v_ashrrev_i16_e64_dpp v5, v1, v2 dpp8:[7,6,5,4,3,2,1,0] ; encoding: [0x05,0x00,0x3a,0xd7,0xe9,0x04,0x02,0x00,0x01,0x77,0x39,0x05]
+0x05,0x10,0x3a,0xd7,0xe9,0x04,0x02,0x00,0x01,0x77,0x39,0x05
+
+# W32-REAL16: v_ashrrev_i16_e64_dpp v255.h, v255.l, v255.l op_sel:[0,0,1] dpp8:[0,0,0,0,0,0,0,0] fi:1 ; encoding: [0xff,0x40,0x3a,0xd7,0xea,0xfe,0x03,0x00,0xff,0x00,0x00,0x00]
+# W32-FAKE16: v_ashrrev_i16_e64_dpp v255, v255, v255 dpp8:[0,0,0,0,0,0,0,0] fi:1 ; encoding: [0xff,0x00,0x3a,0xd7,0xea,0xfe,0x03,0x00,0xff,0x00,0x00,0x00]
+# W64-REAL16: v_ashrrev_i16_e64_dpp v255.h, v255.l, v255.l op_sel:[0,0,1] dpp8:[0,0,0,0,0,0,0,0] fi:1 ; encoding: [0xff,0x40,0x3a,0xd7,0xea,0xfe,0x03,0x00,0xff,0x00,0x00,0x00]
+# W64-FAKE16: v_ashrrev_i16_e64_dpp v255, v255, v255 dpp8:[0,0,0,0,0,0,0,0] fi:1 ; encoding: [0xff,0x00,0x3a,0xd7,0xea,0xfe,0x03,0x00,0xff,0x00,0x00,0x00]
+0xff,0x40,0x3a,0xd7,0xea,0xfe,0x03,0x00,0xff,0x00,0x00,0x00
+
+# GFX11: v_ashrrev_i16_e64_dpp v5.h, v1.h, v2.h op_sel:[1,1,1] dpp8:[7,6,5,4,3,2,1,0] ; encoding: [0x05,0x58,0x3a,0xd7,0xe9,0x04,0x02,0x00,0x01,0x77,0x39,0x05]
```
Comment (Sisyph) [inline llvm/test/MC/Disassembler/AMDGPU/gfx12_dasm_vop3_dpp8.txt:272]: Stray GFX11 check in a gfx12 file

File: llvm/test/MC/Disassembler/AMDGPU/gfx12_dasm_vop3_dpp16.txt [hunk 4]
```diff
@@ -413,48 +413,142 @@
...
+0xff,0x40,0x3a,0xd7,0xfa,0xfe,0x03,0x00,0xff,0x6f,0x0d,0x30
+
+# GFX11: v_ashrrev_i16_e64_dpp v5.h, v1.h, v2.h op_sel:[1,1,1] row_share:0 row_mask:0xf bank_mask:0xf ; encoding: [0x05,0x58,0x3a,0xd7,0xfa,0x04,0x02,0x00,0x01,0x50,0x01,0xff]
+# W32-REAL16: v_ashrrev_i16_e64_dpp v5.h, v1.h, v2.h op_sel:[1,1,1] row_share:0 row_mask:0xf bank_mask:0xf ; encoding: [0x05,0x58,0x3a,0xd7,0xfa,0x04,0x02,0x00,0x01,0x50,0x01,0xff]
+# W32-FAKE16: v_ashrrev_i16_e64_dpp v5, v1, v2 row_share:0 row_mask:0xf bank_mask:0xf ; encoding: [0x05,0x00,0x3a,0xd7,0xfa,0x04,0x02,0x00,0x01,0x50,0x01,0xff]
+# W64-REAL16: v_ashrrev_i16_e64_dpp v5.h, v1.h, v2.h op_sel:[1,1,1] row_share:0 row_mask:0xf bank_mask:0xf ; encoding: [0x05,0x58,0x3a,0xd7,0xfa,0x04,0x02,0x00,0x01,0x50,0x01,0xff]
+# W64-FAKE16: v_ashrrev_i16_e64_dpp v5, v1, v2 row_share:0 row_mask:0xf bank_mask:0xf ; encoding: [0x05,0x00,0x3a,0xd7,0xfa,0x04,0x02,0x00,0x01,0x50,0x01,0xff]
+0x05,0x58,0x3a,0xd7,0xfa,0x04,0x02,0x00,0x01,0x50,0x01,0xff
+
+# GFX11: v_ashrrev_i16_e64_dpp v5.l, v1.h, v2.l op_sel:[1,0,0] row_share:15 row_mask:0x0 bank_mask:0x1 ; encoding: [0x05,0x08,0x3a,0xd7,0xfa,0x04,0x02,0x00,0x01,0x5f,0x01,0x01]
```
Comment (Sisyph) [inline llvm/test/MC/Disassembler/AMDGPU/gfx12_dasm_vop3_dpp16.txt:531]: Stray gfx11 check

File: llvm/test/MC/Disassembler/AMDGPU/gfx12_dasm_vop3.txt [hunk 5]
```diff
@@ -513,51 +513,117 @@
...
+0x05,0x08,0x3a,0xd7,0x01,0x05,0x02,0x00
+
+# GFX11: v_ashrrev_i16 v5.l, v255.l, v255.h op_sel:[0,1,0] ; encoding: [0x05,0x10,0x3a,0xd7,0xff,0xff,0x03,0x00]
+# W32-REAL16: v_ashrrev_i16 v5.l, v255.l, v255.h op_sel:[0,1,0] ; encoding: [0x05,0x10,0x3a,0xd7,0xff,0xff,0x03,0x00]
+# W32-FAKE16: v_ashrrev_i16 v5, v255, v255            ; encoding: [0x05,0x00,0x3a,0xd7,0xff,0xff,0x03,0x00]
+# W64-REAL16: v_ashrrev_i16 v5.l, v255.l, v255.h op_sel:[0,1,0] ; encoding: [0x05,0x10,0x3a,0xd7,0xff,0xff,0x03,0x00]
+# W64-FAKE16: v_ashrrev_i16 v5, v255, v255            ; encoding: [0x05,0x00,0x3a,0xd7,0xff,0xff,0x03,0x00]
+0x05,0x10,0x3a,0xd7,0xff,0xff,0x03,0x00
+
+# GFX11: v_ashrrev_i16 v255.h, 0xfe0b, vcc_hi op_sel:[0,0,1] ; encoding: [0xff,0x40,0x3a,0xd7,0xff,0xd6,0x00,0x00,0x0b,0xfe,0x00,0x00]
```
Comment (Sisyph) [inline llvm/test/MC/Disassembler/AMDGPU/gfx12_dasm_vop3.txt:620]: Misplaced GFX11

File: llvm/test/MC/Disassembler/AMDGPU/gfx12_dasm_vop3_dpp8.txt (hunk 3 above)
Comment (broxigarchen) [reply llvm/test/MC/Disassembler/AMDGPU/gfx12_dasm_vop3_dpp8.txt]: removed all

**Review by Sisyph — APPROVED** [review]
LGTM

**Pull request merged**